### PR TITLE
Redesigned clip and scroll API

### PR DESCRIPTION
**Pull request by kvark**
Fixes https://bugzilla.mozilla.org/show_bug.cgi?id=1503447
Note: this description is edited to contain the relevant up-to-date information about the code.

## General idea
This is something I wanted for a long time: the spatial hierarchy that isn't mixed up with the clip hierarchy. The change dragged a bunch of other API fixes that were accumulated in our technical debt, so it grew rather big/scary.

The motivation is - removing the complexity of the API that we inherited from the past. Both Servo and Gecko are in a better position to explicitly tell us what the clip/spatial properties of items are, and they don't need to mix them up in the same type. The old API had more complex logic, more state, and caused the Gecko integration to be painful and non-solid (there are edge cases not handled properly).

## API changes:
 - separate `SpatialId` from `ClipId` :tada: 
 - remove the clip/scroll stack from the API, all the `SpatialId` and `ClipId` are passed explicitly for display items in the form of `SpaceAndClipInfo`, which roughly corresponds to the old `ClipAndScrollInfo`.
  - ~~reference frame are no longer pushed/popped and instead are "defined", similar to scroll/sticky frames.~~ Edit: this ended up [harder than I expected](https://phabricator.services.mozilla.com/D13081), postponing until the main change is done.
  - stacking contexts no longer have an internal `ClipId` passed in (and the standard one being ignored), instead they use the standard one from the associated `SpaceAndClipInfo`.
  - `define_clip_with_parent` and `define_scroll_frame_with_parent` are removed.

## TODO:
- [ ] reviews:
  - [ ] here
  - [ ] Gecko part
  - [ ] Servo part
- [x] integrate with Gecko and provide a good try push:
  - https://treeherder.mozilla.org/#/jobs?repo=try&selectedJob=218192168&revision=500bb4429beb003927afb5e14201587519dce462
- [x] integrate with Servo:
  - https://github.com/kvark/servo/commit/f671511a71f7dc61eab4c4342c659d928d561888 
  - there are failures, but they are exactly the same as without this change

---
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/3251)

### CODE REVIEW

**Comment by gw3583** [comment]
It would be good to also provide a patch for Servo as part of this PR, or some detailed instructions on what will be required to change in the next WR update in Servo.

**Comment by gw3583** [comment]
(I haven't reviewed this yet, btw - just an initial comment above)

**Comment by gw3583** [comment]
@mrobinson Thanks for jumping in to help with reviewing this! :+1:

**Comment by kvark** [comment]
@mrobinson it's a pleasant surprise to see you, thank for taking a look!

File: webrender/src/display_list_flattener.rs [hunk 1]
```diff
@@ -250,11 +235,9 @@ impl<'a> DisplayListFlattener<'a> {
 
     fn flatten_root(&mut self, pipeline: &'a ScenePipeline, frame_size: &LayoutSize) {
         let pipeline_id = pipeline.pipeline_id;
-        let reference_frame_info = self.simple_scroll_and_clip_chain(
-            &ClipId::root_reference_frame(pipeline_id),
-        );
+        let root_scroll_node = SpatialId::root_scroll_node(pipeline_id);
 
-        let root_scroll_node = ClipId::root_scroll_node(pipeline_id);
+        let reference_frame_info = self.simple_scroll_and_clip_chain(&root_scroll_node);
```
[the remark this answers is not in the document]
Comment (kvark) [reply webrender/src/display_list_flattener.rs]: nicely spotted! This is a mistake

File: webrender/src/display_list_flattener.rs [hunk 2]
```diff
@@ -1343,15 +1318,13 @@ impl<'a> DisplayListFlattener<'a> {
...
 
         // Map from parent ClipId to existing clip-chain.
-        let mut parent_clip_chain_index = self
-            .id_to_index_mapper
-            .get_clip_chain_id(&parent_id);
-        // Map the ClipId for the positioning node to a spatial node index.
-        let spatial_node = self.id_to_index_mapper.get_spatial_node_index(parent_id);
+        let mut parent_clip_chain_index = match cs_info.clip_node_id {
+            Some(clip_id) => self.id_to_index_mapper.get_clip_chain_id(clip_id),
+            None => ClipChainId::NONE,
```
[the remark this answers is not in the document]
Comment (kvark) [reply webrender/src/display_list_flattener.rs]: This piece is not a part of the API, it's the DL processing. At this level, we just strictly respect the parent clip chain specified. It's up to the `webrender_api/src/display_list.rs` to ensure that the "implied" parent semantics is preserved.

I'm going to test Gecko shortly and see what the implications are. I'd expect wrench reftests to fail if this semantics was ported incorrectly.

File: webrender_api/src/display_item.rs [hunk 3]
```diff
@@ -22,28 +22,24 @@ pub const MAX_BLUR_RADIUS: f32 = 300.;
...
 #[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
 pub struct ClipAndScrollInfo {
-    pub scroll_node_id: ClipId,
+    pub scroll_node_id: SpatialId,
     pub clip_node_id: Option<ClipId>,
 }
 
 impl ClipAndScrollInfo {
-    pub fn simple(node_id: ClipId) -> ClipAndScrollInfo {
+    pub fn simple(scroll_node_id: SpatialId) -> Self {
```
[the remark this answers is not in the document]
Comment (kvark) [reply webrender_api/src/display_item.rs]: I removed `ClipAndScrollInfo` altogether now. No need to group together loosely related things.

**Comment by kvark** [comment]
Asking for a real review now, while I'm finishing testing Gecko side.

**Comment by gw3583** [comment]
Just one question above, since it's not clear to me that this should have affected any reftest outputs (and if that has any implications for Gecko / Servo).

Otherwise, it seems like we need:
 - [ ] Try run
 - [ ] Finish up any remaining issues from @mrobinson 's comments.
 - [ ] Get sign off from @staktrace 

Then this should be ready to merge.

**Comment by kvark** [comment]
@gw3583 I'm working on even more changes now that remove the clip-scroll stack completely from the API. More updates to come... and I'll double check the test. I assume there is a slight change in YAML reading semantics (e.g. parsing the "clip-and-scroll" field), since it diverges further from the actual WR API, and I'm trying to preserve compatibility with the existing reftests.

**Comment by kvark** [comment]
I've got to document this somewhere: we have a few places in DL building where the clip is literally smuggled through the scroll ID, while the clip ID is silently dropped:
  - `push_reference_frame`
  - `define_clip`
  - possibly other

What happens is `scroll_node_id` gets through, and the DL flattener associates the produced clip (for `define_clip`) or scroll (for `push_reference_frame`) with the clip chain of the parent, hence the clipping survives even though the original `clip_node_id` is ignored.

But the worst thing is that there is code in Gecko that relies on that behavior...

**Comment by kvark** [comment]
@gw3583 looks like the try is green: https://treeherder.mozilla.org/#/jobs?repo=try&selectedJob=218192168&revision=500bb4429beb003927afb5e14201587519dce462

**Comment by kvark** [comment]
Rebased on latest try: https://treeherder.mozilla.org/#/jobs?repo=try&selectedJob=218226683&revision=73a1fb8e56d4e5f9580af422c5c9892923f8afde

File: webrender/src/display_list_flattener.rs [hunk 4]
```diff
@@ -1766,7 +1746,8 @@ impl<'a> DisplayListFlattener<'a> {
             frame_kind,
         );
         self.id_to_index_mapper.map_spatial_node(new_node_id, node_index);
-        self.id_to_index_mapper.map_to_parent_clip_chain(new_node_id, &parent_id);
+        //TODO: remove
```
Comment (staktrace) [inline webrender/src/display_list_flattener.rs]: drop this if unneeded

**Comment by kvark** [comment]
@staktrace I think those `//TODO` comments in the intermediate commits are fine to keep. After all, these aren't *fully* safe contained changes. I made sure each one compiles, but reftests are only good for the full PR, not each change. I can squash the whole thing if you want the commits to be reftestable independently.

File: webrender/src/display_list_flattener.rs [hunk 5]
```diff
@@ -582,33 +564,34 @@ impl<'a> DisplayListFlattener<'a> {
...
+
     fn flatten_stacking_context(
         &mut self,
         traversal: &mut BuiltDisplayListIter<'a>,
         pipeline_id: PipelineId,
-        item: &DisplayItemRef,
         stacking_context: &StackingContext,
-        scroll_node_id: ClipId,
-        reference_frame_relative_offset: LayoutVector2D,
+        origin: LayoutPoint,
```
Comment (staktrace) [inline webrender/src/display_list_flattener.rs]: Not really clear to me why you chose to replace the `item: &DisplayItemRef` with `origin: LayoutPoint, filters: ItemRange<FilterOp>` here but you didn't do the equivalent thing in e.g. `flatten_iframe`. Doesn't matter much functionally, but this is the sort of unrelated change that slows down my reviews...

**Comment by kvark** [comment]
New try with everything rebased: https://treeherder.mozilla.org/#/jobs?repo=try&revision=fb7b18e47f87a7c551764d96e3e2a92e13f83747

**Comment by staktrace** [comment]
> @staktrace I think those `//TODO` comments in the intermediate commits are fine to keep.

I think they got rebased away anyway, so it's a moot point

> After all, these aren't _fully_ safe contained changes. I made sure each one compiles, but reftests are only good for the full PR, not each change. I can squash the whole thing if you want the commits to be reftestable independently.

No, I think it's reasonable that reftests only pass with the full set, but that each patch compiles.

**Comment by staktrace** [comment]
Looks like the latest force-push just removed a spurious logfile so the comments I made above are still relevant. Will continue reviewing where I left off.

**Comment by kvark** [comment]
@staktrace thank you for reviewing! I'll stop force-pushing for now :) Although it will be needed to resolve the merge conflict now.

File: webrender_api/src/display_list.rs [hunk 6]
```diff
@@ -1455,16 +1455,13 @@ impl DisplayListBuilder {
         layout: &LayoutPrimitiveInfo,
         space_and_clip: &SpaceAndClipInfo,
         pipeline_id: PipelineId,
-        ignore_missing_pipeline: bool,
-    ) -> ClipId {
-        let id = self.generate_clip_index();
+        ignore_missing_pipeline: bool
```
Comment (staktrace) [inline webrender_api/src/display_list.rs]: nit: dropped a comma here

File: examples/basic.rs [hunk 7]
```diff
@@ -196,7 +196,8 @@ impl Example for App {
 
         builder.push_stacking_context(
             &info,
-            &root_space_and_clip,
+            root_space_and_clip.spatial_id,
```
Comment (staktrace) [inline examples/basic.rs]: Could just use `spatial_id` here since that's defined a few lines up

File: wrench/src/yaml_frame_writer.rs [hunk 8]
```diff
@@ -1247,43 +1246,61 @@ impl webrender::ApiRecordingReceiver for YamlFrameWriterReceiver {
...
         } else if id.is_root_scroll_node() {
             Yaml::String("root-scroll-node".to_owned())
         } else {
-            Yaml::Integer(*self.hash_map.get(id).unwrap() as i64)
+            Yaml::Integer(self.spatial_map[id] as i64)
+        }
+    }
+
+    fn map_clip_id(&self, id: &ClipId) -> Yaml {
+        if id.is_root() {
```
[the remark this answers is not in the document]
Comment (kvark) [reply wrench/src/yaml_frame_writer.rs]: could you clarify this one, please?

File: wrench/src/yaml_frame_reader.rs [hunk 9]
```diff
@@ -362,50 +384,83 @@ impl YamlFrameReader {
...
+            Yaml::Integer(value) => self.user_spatial_id_map[&(value as u64)],
             Yaml::String(ref id_string) if id_string == "root-reference-frame" =>
-                Some(ClipId::root_reference_frame(pipeline_id)),
+                SpatialId::root_reference_frame(pipeline_id),
             Yaml::String(ref id_string) if id_string == "root-scroll-node" =>
-                Some(ClipId::root_scroll_node(pipeline_id)),
-            _ => None,
+                SpatialId::root_scroll_node(pipeline_id),
+            _ => {
+                println!("Unable to parse SpatialId {:?}", item);
```
[the remark this answers is not in the document]
Comment (kvark) [reply wrench/src/yaml_frame_reader.rs]: Currently this module uses `println!` for ignorable errors/warnings, so I'm just continuing the trend.

File: wrench/src/rawtest.rs [hunk 10]
```diff
@@ -1025,30 +1080,31 @@ impl<'a> RawtestHarness<'a> {
...
-        builder.push_rect(&info, ColorF::new(1.0, 1.0, 1.0, 1.0));
-        builder.pop_clip_id();
-
+        builder.push_rect(&info, &space_and_clip, ColorF::new(1.0, 1.0, 1.0, 1.0));
 
         // Add a rectangle that is clipped by a ClipChain containing a rounded rect.
         let rect = LayoutRect::new(LayoutPoint::new(200., 100.), LayoutSize::new(100., 100.));
-        let clip_id = builder.define_clip(rect, vec![make_rounded_complex_clip(&rect, 20.)], None);
+        let clip_id = builder.define_clip(
+            &space_and_clip,
```
[the remark this answers is not in the document]
Comment (kvark) [reply wrench/src/rawtest.rs]: wow, amazing find! :clap: :medal_military:

File: wrench/src/yaml_frame_writer.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (staktrace) [reply wrench/src/yaml_frame_writer.rs]: I was thinking that if this passed a `ClipId` for which `is_valid()` returns false, you probably want to serialize it as `Yaml::String("invalid".to_owned())`. Although I don't think it should ever actually happen, so maybe just asserting is good enough.

[the remark this answers is not in the document]
Comment (kvark) [reply wrench/src/yaml_frame_writer.rs]: I agree that having a special "invalid" would improve things, but this seems to be slightly out of scope for the PR (i.e. we already have this problem)

[the remark this answers is not in the document]
Comment (staktrace) [reply wrench/src/yaml_frame_writer.rs]: But this patch is the one that adds the concept of an "invalid" clip id, it didn't exist prior to this.

File: webrender/src/display_list_flattener.rs (hunk 5 above)
Comment (kvark) [reply webrender/src/display_list_flattener.rs]: I figured that the internal methods shouldn't be working with `DisplayItemRef` and instead just receive the data they need, but ended up only fixing the spots I was working with. It was already not very consistent there, so I didn't make it worse. I think we should just follow-up with this later.

File: wrench/src/yaml_frame_writer.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (kvark) [reply wrench/src/yaml_frame_writer.rs]: oh snap, I've been looking at it for so long that I mix the past with the future now

**Comment by kvark** [comment]
@staktrace alright, *now* everything is addressed :)

**Comment by kvark** [comment]
@gw3583 @jrmuizel Here is the latest up-to-date status:

Code is considered ready by me and the reviewers (thanks @staktrace !). Gecko side: https://phabricator.services.mozilla.com/D15230

Rebased try push shows a number of unexpected passes and 2 failures in Windows/Debug: https://treeherder.mozilla.org/#/jobs?repo=try&selectedJob=219135236&revision=a0bc8595ca76b1cae0e53220bfbf01550f185fba
I find this quite odd, since it's only in debug, and I didn't see those before the rebase. Launched a try from the same FF revision without my changes to compare: https://treeherder.mozilla.org/#/jobs?repo=try&selectedJob=219135236&revision=40d93ef7d636034ff37e7c9aa79762437b099055

@darkspirit gave it a short run on Linux, didn't find any issues, was going to also run on Win10 today and I hope ping back.

These are the only issues to date. If all is good by the end of the day, we may proceed with landing.

**Comment by staktrace** [comment]
>  Rebased try push shows a number of unexpected passes and 2 failures in Windows/Debug: https://treeherder.mozilla.org/#/jobs?repo=try&selectedJob=219135236&revision=a0bc8595ca76b1cae0e53220bfbf01550f185fba I find this quite odd, since it's only in debug, and I didn't see those before the rebase.

I retriggered the job a few more times and it looks like that was just a random intermittent. The retriggers are all green so far.

**Comment by Darkspirit** [comment]
LGTM

**Comment by staktrace** [comment]
Let's do this

@bors-servo r+
